src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs

```diff
@@ -26,6 +27,11 @@ private enum SerializationMask
         private static Func<IEnumerable<ClaimsIdentity>, ClaimsIdentity?> s_identitySelector = SelectPrimaryIdentity;
         private static Func<ClaimsPrincipal> s_principalSelector = ClaimsPrincipalSelector;
 
+        private static ClaimsPrincipal SelectClaimsPrincipal()
+        {
+            return (Thread.CurrentPrincipal is ClaimsPrincipal claimsPrincipal) ? claimsPrincipal : new ClaimsPrincipal(Thread.CurrentPrincipal!);
```

Comment:
`Thread.CurrentPrincipal` can be null (thus the reason you are using `!` here):

This code will make `Thread.CurrentPrincipal` return null:
```C#
AppDomain.CurrentDomain.SetPrincipalPolicy(PrincipalPolicy.NoPrincipal);
```

And `new ClaimsPrincipal(null)` will throw an `ArgumentNullException`, which I don't think we want from the static `ClaimsPrincipal.Current` property.

Reply:
That looks to be the exact behavior that occurred on .NET Framework: <https://referencesource.microsoft.com/#mscorlib/system/security/claims/ClaimsPrincipal.cs,458c87e88a4dcaee>

Reply:
1. That seems like very odd behavior, and sounds like a bug to me. Are we sure we want to duplicate that behavior?
2. `public static ClaimsPrincipal? Current` is already annotated as nullable, why not return `null` in this case?
3. If this is the behavior we want (or even if we want different behavior), it seems like a good test to write.

Reply:
> That seems like very odd behavior, and sounds like a bug to me. Are we sure we want to duplicate that behavior?

Yep.  There's not really investment in this area, and it has no clear owner (nominally, I own it, but I don't know anything about it, or do anything with it, and dread any issues/PRs for it).  So we just want it to be just like NetFX, barring strong data otherwise.

> public static ClaimsPrincipal? Current is already annotated as nullable, why not return null in this case?

Because NetFX doesn't.

> If this is the behavior we want (or even if we want different behavior), it seems like a good test to write.

I vaguely agree; but if it's just matching NetFX, and no one understands it, I don't have strong opinions.  I'd rather get it matching-with-no-additional-tests than miss the release; since this is in response to someone actually observing the difference.

Reply:
> I'd rather get it matching-with-no-additional-tests than miss the release; since this is in response to someone actually observing the difference.

Since someone actually observed the difference, wouldn't we want to ensure the behavior in fact matches and continues to match in the future?

Reply:
At this point, I believe @tannergooding is on vacation until it's too late to do anything about it; and I'm buried in other projects.  If you want to contribute a test to this PR, it'll definitely make me smile.

Reply:
If you want, you can take the PR as-is, open an issue to follow up with a test and assign it to me. I should be able to get to it before Friday morning.

Reply:
Thanks for the offer, @vcsjones! Since @tannergooding is on vacation I wrote a quick test for this at @bartonjs's suggestion. PTAL.